File: app/src/main/java/com/ocram/qichwadic/data/datasource/SearchCloudDataStore.java
package com.ocram.qichwadic.data.datasource;

import com.ocram.qichwadic.data.repository.SearchRepositoryImpl;
import com.ocram.qichwadic.domain.model.Definition;
import com.ocram.qichwadic.domain.model.SearchResult;
import com.ocram.qichwadic.framework.net.client.RetrofitClient;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Flowable;

public class SearchCloudDataStore implements SearchRepositoryImpl.SearchCloudDataStore {

    private final RetrofitClient retrofitClient;

    @Inject
    public SearchCloudDataStore(RetrofitClient retrofitClient) {
        this.retrofitClient = retrofitClient;
    }

    @Override
    public Flowable<List<SearchResult>> search(int fromQuechua,String target, String word, int searchType) {
        return retrofitClient.searchWord(word, fromQuechua, target, searchType);
    }

    @Override
    public Flowable<List<Definition>> fetchMoreResults(int dictionaryId, String word, int searchType, int page) {
        return retrofitClient.fetchMoreResults(dictionaryId, word, 0, searchType, page);
    }
}

File: app/src/main/java/com/ocram/qichwadic/features/search/data/datastore/SearchCloudDataStore.kt
package com.ocram.qichwadic.features.search.data.datastore

import com.ocram.qichwadic.core.data.remote.ApiResponse
import com.ocram.qichwadic.core.data.remote.CloudDataStore
import com.ocram.qichwadic.core.data.model.SearchResultEntity
import com.ocram.qichwadic.core.data.remote.RetrofitClient

interface SearchCloudDataStore {
    suspend fun search(fromQuechua: Int, target: String, word: String, searchType: Int): ApiResponse<out List<SearchResultEntity>>

    suspend fun fetchMoreResults(dictionaryId: Int, word: String, searchType: Int, page: Int): ApiResponse<out SearchResultEntity>
}

class SearchCloudDataStoreImpl(private val retrofitClient: RetrofitClient): CloudDataStore(), SearchCloudDataStore {

    override suspend fun search(fromQuechua: Int, target: String, word: String, searchType: Int): ApiResponse<out List<SearchResultEntity>> {
        return processResponse(retrofitClient.searchWord(word, fromQuechua, target, searchType))
    }

    override suspend fun fetchMoreResults(dictionaryId: Int, word: String, searchType: Int, page: Int): ApiResponse<out SearchResultEntity> {
        return processResponse(retrofitClient.fetchMoreResults(dictionaryId, word, 0, searchType, page))
    }
}
